### PR TITLE
Refactor/provide option to return results prior to errors

### PR DESCRIPTION
**Pull request by em-gazelle**
- Added new class of StandardError to capture params and results fetched from previous pages in pagination loop

### CODE REVIEW

File: lib/booker/errors/mid_pagination_error.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+class MidPaginationError
```
Comment (mstahl) [inline lib/booker/errors/mid_pagination_error.rb]: Probably don't need this if we're using the MidPaginationError above.

File: spec/lib/booker/errors_spec.rb [hunk 2]
```diff
@@ -69,6 +69,42 @@
...
+  describe '.new' do
+    let(:error) { Booker::MidPaginationError.new(message, error_occurred_during_params, results_fetched_prior_to_error) }
+
+    before do
+      expect(error.error_occurred_during_params).to eq error_occurred_during_params
+      expect(error.results_fetched_prior_to_error).to eq results_fetched_prior_to_error
+      expect(error.message).to eq message
+    end
+
+    context 'when all atRtrs defined' do
```
Comment (mstahl) [inline spec/lib/booker/errors_spec.rb]: Typo

File: spec/lib/booker/errors_spec.rb [hunk 3]
```diff
@@ -69,6 +69,42 @@
...
+      let(:error_occurred_during_params) { {PageSize: 1} }
+      let(:results_fetched_prior_to_error) { ['prior results'] }
+      let(:message) { "Special message" }
+
+      it 'sets error data' do; end
```
Comment (mstahl) [inline spec/lib/booker/errors_spec.rb:99]: Probably should have a test here.

File: spec/lib/booker/errors_spec.rb [hunk 4]
```diff
@@ -69,6 +69,42 @@
...
+    end
+    context 'when attrs undefined' do
+      let(:error_occurred_during_params) { {} }
+      let(:results_fetched_prior_to_error) { [] }
+      let(:message) { "Error occurred during call mid-pagination" }
+
+      it 'sets defaults' do; end
```
Comment (mstahl) [inline spec/lib/booker/errors_spec.rb:105]: Here too.

File: spec/lib/booker/errors_spec.rb [hunk 5]
```diff
@@ -69,6 +69,41 @@
...
+    it { expect(subject).to respond_to :message= }
+  end
+
+  describe '.new' do
+    let(:error) { Booker::MidPaginationError.new(message, error_occurred_during_params, results_fetched_prior_to_error) }
+    let(:message) { "Error occurred during call mid-pagination" }
+    before do
+      expect(error.error_occurred_during_params).to eq error_occurred_during_params
+      expect(error.results_fetched_prior_to_error).to eq results_fetched_prior_to_error
+      expect(error.message).to eq message
```
Comment (em-gazelle) [inline spec/lib/booker/errors_spec.rb:91]: I dried these up by putting them in the before block; they're the same for both undefined/defined.

File: spec/lib/booker/errors_spec.rb (hunk 3 above)
Comment (em-gazelle) [reply spec/lib/booker/errors_spec.rb:99]: they're defined/called in the before block; I'm testing that all attributes are set both when they are defined and undefined

**Review by mstahl — APPROVED** [review]
:shipit:

**Pull request merged**